test/albumController.test.js

```diff
@@ -0,0 +1,105 @@
+const { app, connectDB } = require('../src/expressServer'); // Import your Express app here
+const { StatusCodes } = require('http-status-codes');
+const request = require('supertest');
+const { MongoMemoryServer } = require('mongodb-memory-server');
+const Album = require('../src/models/Album');
+
+const {
+  getAllAlbums,
+  getSingleAlbum,
+  getAlbumWithAllUsersWhoPurchasedIt,
+  getFilteredAlbums,
+} = require('../src/controllers/albumController');
+
+let server;
+let mongooseConnection;
+let mongodb;
+
+beforeAll(async () => {
+  mongodb = await MongoMemoryServer.create();
+  const url = mongodb.getUri();
+  process.env.MONGO_URL = url;
+  mongooseConnection = await connectDB(url);
+  server = await app.listen(8001);
+});
+
+afterAll(async () => {
+  await server.close();
+  await mongooseConnection.disconnect();
+  await mongodb.stop();
+});
+
+describe('AlbumController API Tests', () => {
+  const mockAlbumData = {
+    albumName: 'Test Album',
+    artistName: 'Test Artist',
+    price: 9.99,
+    spotifyUrl: 'https://api.spotify.com/v1/albums/',
+  };
+
+  beforeEach(async () => {
+    await Album.deleteMany({});
+    await Album.create(mockAlbumData);
+  });
+
+  it('should test the getAllAlbums endpoint - Success Case', async () => {
+    const response = await request(app).get('/api/v1/albums');
+    expect(response.status).toBe(StatusCodes.OK);
+    expect(response.body).toHaveProperty('albums');
+    expect(response.body.albums).toHaveLength(1);
+  });
+
+  it('should test the getAllAlbums endpoint - Error Case (Unauthorized)', async () => {
+    const response = await request(app).get('/api/v1/albums');
+    expect(response.status).toBe(StatusCodes.UNAUTHORIZED);
+  });
+
+  it('should test the getSingleAlbum endpoint - Success Case', async () => {
+    const albums = await Album.find({});
+    const albumId = albums[0]._id;
+
+    const response = await request(app).get(`/api/albums/v1/${albumId}`);
+    expect(response.status).toBe(StatusCodes.OK);
+    expect(response.body).toHaveProperty('album');
+  });
+
+  it('should test the getSingleAlbum endpoint - Error Case (Not Found)', async () => {
+    const response = await request(app).get('/api/albums/v1/invalidAlbumId');
+    expect(response.status).toBe(StatusCodes.NOT_FOUND);
+  });
+
+  it('should test the getAlbumWithAllUsersWhoPurchasedIt endpoint - Success Case', async () => {
+    const albums = await Album.find({});
+    const albumId = albums[0]._id;
+
+    const response = await request(app).get(
+      `/api/albums/${albumId}/purchased-users`
+    );
+    expect(response.status).toBe(StatusCodes.OK);
+    expect(response.body).toHaveProperty('album');
```

Comment:
Since this test is specifically about making sure this endpoint returns the album _**with**_ all the users who purchased it; it's probably worthwhile to also assert that in our expectations here:

```suggestion
    expect(response.body).toHaveProperty('album');
    expect(response.body).toHaveProperty('album');
```

Reply:
Oops in my previous comment I didn't include what I meant by that 😮‍💨 ; this endpoint returns an `album` object with an additional field that shows all the users who purchased it. We should add at least one `expect` assertion here to verify that the response body we get actually has that field; this will help us catch any bugs or errors in the future if the shape of this response is accidentally changed in the controller:

```json
    {
      album: {
        _id: '650dc79709eee6395154e250',
        artistName: 'Test Artist',
        albumName: 'Test Album',
        price: 9.99,
        image: '/uploads/example.jpg',
        spotifyUrl: 'https://api.spotify.com/v1/albums/blah123',
        averageRating: 0,
        numOfReviews: 0,
        createdAt: '2023-09-22T16:57:59.195Z',
        updatedAt: '2023-09-22T16:57:59.195Z',
        __v: 0,
        purchasedByUsers: [],
        id: '650dc79709eee6395154e250'
      },
      purchasingUsersCount: 0
    }
```